### PR TITLE
[Refactor] Change to BNC data flow for all DETRs

### PR DESCRIPTION
**Pull request by KeiChiTse**
## Motivation
This PR unifies the data flow to BNC for all DETRs, where B denotes batch size, N denotes sequence length, C denotes embedding dimension.

I'm working together with @jshilong, @Li-Qingyun, and @LYMDLUT to refactor the DETR-like algorithms in MMDetection 3.x, using a more lightweight and easy-to-read code style.

Details can be refered to in brother PRs #9252, #9149, #8362, #8763, #9248, feel free to contact me and give your advice!

### CODE REVIEW

File: mmdet/models/layers/transformer/detr_transformer.py [hunk 1]
```diff
@@ -263,16 +252,10 @@ def __init__(self,
                      act_cfg=dict(type='ReLU', inplace=True),
                  ),
                  norm_cfg: OptConfigType = dict(type='LN'),
-                 init_cfg: OptConfigType = None,
-                 batch_first: bool = False) -> None:
+                 init_cfg: OptConfigType = None) -> None:
```
Comment (jshilong) [inline mmdet/models/layers/transformer/detr_transformer.py:255]:
```
if "batch_first" not in  self.self_attn_cfg:
    self.self_attn_cfg["batch_first"] = True
else:
   assert self.self_attn_cfg["batch_first"] == True, "First dimension of all DETRs in mmdet is `batch`, please set `batch_first`  flag"
```

Comment (jshilong) [reply mmdet/models/layers/transformer/detr_transformer.py:255]: please add this assert for all configs for pytorch's attention

**Comment by jshilong** [comment]
LGTM, @Li-Qingyun Can you help to infer the checkpoints to make sure the modifications are correct

File: mmdet/models/layers/transformer/detr_transformer.py (hunk 1 above)
Comment (KeiChiTse) [reply mmdet/models/layers/transformer/detr_transformer.py:255]: > 

Should we set `self.cross_attn_cfg["batch_first"] = True`? After all, `MultiheadAttention` and `MultiScaleDeformableAttention` both need the `batch_first` flag.

Comment (KeiChiTse) [reply mmdet/models/layers/transformer/detr_transformer.py:255]: > s

ok

Comment (KeiChiTse) [reply mmdet/models/layers/transformer/detr_transformer.py:255]: > please add this assert for all configs for pytorch's attention

Maybe not only pytorch's attention? MultiScaleDeformableAttention needs this assert too.

**Comment by jshilong** [comment]
LGTM

**Comment by Li-Qingyun** [comment]
> LGTM, @Li-Qingyun Can you help to infer the checkpoints to make sure the modifications are correct

OK, DETR, DeformDETR, DeformDETR+, DeformDETR++ have been aligned.

图片
图片

图片

图片

File: mmdet/models/dense_heads/detr_head.py [hunk 2]
```diff
@@ -132,29 +132,30 @@ def _init_layers(self) -> None:
             add_residual=False)
         # NOTE the activations of reg_branch here is the same as
         # those in transformer, but they are actually different
-        # in DAB DETR (prelu in transformer and relu in reg_branch)
+        # in DAB-DETR (prelu in transformer and relu in reg_branch)
         self.fc_reg = Linear(self.embed_dims, 4)
 
     # Note function _load_from_state_dict is deleted without
-    # supporting refactor-DETR in mmdetection2.0
+    # supporting refactor-DETR in mmdetection 2.x
```
Comment (Li-Qingyun) [inline mmdet/models/dense_heads/detr_head.py:139]: I can not remember. what does this mean.

Comment (jshilong) [reply mmdet/models/dense_heads/detr_head.py:139]: just remove it

**Pull request merged**